File: ios/Sources/ClipfireiOS/Models/CompositionTranscript.swift
import Foundation

// MARK: - Stitched-composition transcript helpers
//
// Swift mirror of `polemicyst.com/shared/lib/composition-transcript.ts`.
// Builds the concatenated transcript of a stitched Composition for AI
// title/caption generation.
//
// Why this exists: a Composition is a stitched video — creator clip plus N
// reference tracks, each with their own per-segment transcript. The post-render
// `CompositionOutput.transcript` field is null at the moment the iOS publish
// sheet auto-fires the AI metadata generation (client-side renders never
// populate it; server-side renders enqueue a separate transcription job that
// lands later). Passing an empty transcript to the LLM gives generic,
// content-disconnected titles + captions — the bug this module fixes.
//
// Order is creator transcript first (the spine of the stitch), then each
// reference track sorted by `sortOrder` ascending. Sources are joined with a
// blank line so the LLM sees them as distinct passages — matches the web
// helper exactly so the LLM input is identical between platforms.

/// One segment of a transcript (Whisper / YouTube caption row). Tolerant of
/// extra fields (start/end/etc) returned by the API — only `text` is used here.
public struct TranscriptSegment: Codable, Equatable {
    public let text: String?

    public init(text: String?) {
        self.text = text
    }
}

public enum CompositionTranscript {
    /// Flatten an array of `{ text }` segments into a single string of joined
    /// words. Trims each segment and drops empties so we don't produce double
    /// spaces.
    public static func flattenSegments(_ segments: [TranscriptSegment]?) -> String {
        guard let segments, !segments.isEmpty else { return "" }
        return segments
            .compactMap { $0.text?.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
            .joined(separator: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    /// Build the full stitched-video transcript for an AI prompt. Returns
    /// `nil` if no source has a transcript yet — callers pass `nil` so the
    /// prompt omits the transcript block entirely instead of sending an empty
    /// string.
    ///
    /// Prefers `fallback` (the rendered-output transcript, which captures the
    /// actual stitched audio mix) when truthy. Otherwise concatenates the
    /// creator + per-track transcripts in `sortOrder` ascending.
    public static func buildStitched(
        composition: Composition?,
        fallback: String? = nil
    ) -> String? {
        let fallbackTrimmed = (fallback ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        if !fallbackTrimmed.isEmpty { return fallbackTrimmed }

        guard let composition else { return nil }

        var parts: [String] = []

        let creator = flattenSegments(composition.creatorTranscriptJson)
        if !creator.isEmpty { parts.append(creator) }

        let sortedTracks = (composition.tracks ?? [])
            .sorted { $0.sortOrder < $1.sortOrder }

        for track in sortedTracks {
            let segText = flattenSegments(track.transcriptJson)
            if !segText.isEmpty { parts.append(segText) }
        }

        if parts.isEmpty { return nil }
        return parts.joined(separator: "\n\n")
    }
}

File: shared/lib/composition-transcript.ts
/**
 * Helpers for building the concatenated ("stitched") transcript of a Composition
 * for AI title/caption/description generation.
 *
 * A Composition is a stitched video — it can include the creator clip plus any
 * number of reference tracks, each with their own per-segment transcript. The
 * post-render output gets transcribed in the background (server-side renders) or
 * not at all (client-side renders), so the final `CompositionOutput.transcript`
 * field is often null or stale at the moment the UI auto-fires the AI metadata
 * generation. Passing an empty transcript to the LLM produces generic, content-
 * disconnected titles + captions — the bug this module fixes.
 *
 * Instead, we concatenate the per-source transcripts we already have on the
 * Composition itself (creator transcript JSON + each track's transcript JSON).
 * That gives the AI the actual spoken content of every clip in the stitch, in
 * source order, so the suggested title/caption reflects what's in the video.
 *
 * Pure / framework-agnostic so it can be unit-tested without React or the DB.
 */

export interface TranscriptSegment {
  text: string;
}

export interface CompositionTrackForTranscript {
  /** Order tracks were added in the stitch. Lower = earlier in playback. */
  sortOrder?: number | null;
  /** Whole-clip transcript segments produced by the transcription worker. */
  transcriptJson?: TranscriptSegment[] | null;
}

export interface CompositionForTranscript {
  /** Transcript of the creator's clip (legacy single-track shape). */
  creatorTranscriptJson?: TranscriptSegment[] | null;
  /** All non-creator tracks attached to the composition. */
  tracks?: CompositionTrackForTranscript[] | null;
}

/**
 * Flatten an array of `{ text }` segments into a single string of joined words.
 * Trims each segment and drops empty ones so we don't produce double spaces.
 */
export function flattenTranscriptSegments(
  segments: TranscriptSegment[] | null | undefined
): string {
  // Defensive: the transcription worker can write malformed JSON to the DB
  // (e.g. a partial Whisper response stored as a string, segments missing
  // `text`, non-object entries). The helper's contract is "treat malformed
  // input as no transcript" — never throw, since this runs in the LLM-prompt
  // assembly hot path and a throw kills AI Suggest for every clip in the stitch.
  if (!Array.isArray(segments) || segments.length === 0) return '';
  return segments
    .map((s) => {
      const text = s != null && typeof s === 'object' ? (s as TranscriptSegment).text : undefined;
      return typeof text === 'string' ? text.trim() : '';
    })
    .filter((t) => t.length > 0)
    .join(' ')
    .trim();
}

/**
 * Build the full stitched-video transcript for an AI prompt. Returns `undefined`
 * if no source has a transcript yet — callers pass `undefined` so the prompt
 * omits the transcript block entirely instead of sending an empty string.
 *
 * Order: creator transcript first (it's the spine of the stitch), then each
 * reference track in `sortOrder` ascending. Sources are joined with a blank
 * line so the LLM sees them as distinct passages.
 *
 * @param composition The composition + tracks (only transcript fields needed)
 * @param fallback Optional pre-built transcript to prefer when truthy — used
 *   to honor the rendered-output transcript (which captures the actual stitched
 *   audio) when it has finished being computed.
 */
export function buildStitchedTranscript(
  composition: CompositionForTranscript | null | undefined,
  fallback?: string | null
): string | undefined {
  // Prefer a real rendered-output transcript when one is available.
  const fallbackTrimmed = (fallback ?? '').trim();
  if (fallbackTrimmed.length > 0) return fallbackTrimmed;

  if (!composition) return undefined;

  const parts: string[] = [];

  const creator = flattenTranscriptSegments(composition.creatorTranscriptJson);
  if (creator.length > 0) parts.push(creator);

  const tracks = (composition.tracks ?? [])
    .slice()
    .sort((a, b) => (a.sortOrder ?? 0) - (b.sortOrder ?? 0));

  for (const track of tracks) {
    const segText = flattenTranscriptSegments(track.transcriptJson);
    if (segText.length > 0) parts.push(segText);
  }

  if (parts.length === 0) return undefined;
  return parts.join('\n\n');
}
